### ISSUE TITLE
Can't submit stats for leaderboard

### ISSUE THREAD

**Issue by AndyDnn** [Microsoft/xbox-live-unity-plugin#256]
Our team has been using the xbox-live-unity-plugin for UWP builds of our game.  Everything is going well, but we cannot for the life of us figure out how to properly submit stats.  It seems like it would be intuitive, but we've had smart people working on it for over a week and we can't seem to solve this simple problem.

We've found several web pages that seem to explain the process such as:  

https://docs.microsoft.com/en-us/windows/uwp/xbox-live/get-started-with-creators/add-stats-and-leaderboards-in-unity

… but all of the pages we've found have parts that are unclear or out of date.

We've been experimenting non-stop, but we haven't successfully saved a stat even once.  At this point we're making projects that consist of a single scene that does nothing but try to set a stat with no luck.  If anyone can provide us with some tips on how to submit a stat, either through code or using the prefabs in the xbox-live-unity-plugin, we'd greatly appreciate it.

We set up our stat (Top Score) in the Developer Center.  We have sign in working properly so we're acquiring a user.  We have an Xbox Live Services prefab in our first scene.  

We've tried dropping an IntegerStat prefab into our scene and manipulating it.  That appears to only affect the variable locally.  

We've tried writing a C# script that does nothing but set a score to 20 (just as a test).  It basically calls  XboxLive.Instance.StatsManager.SetStatisticIntegerData(user, "1", 20) directly after setting user to be the current XboxLiveUser.  

We've tried a lot.  Just not the right thing, apparently.  Any suggestions?

**Comment by rakhad** [comment]
[Here](https://youtu.be/eSJpbQHT8vc?t=827) is a video detailing how to set stats.

If this doesn't work or you're still having issues feel free to reopen this.

**Comment by AndyDnn** [comment]
Thanks for sharing that video (and for making it).  It was very well made.  Unfortunately we've watched it many times and I still cannot get Stats to save.  I've gone so far as to make a minimal sample project that does nothing but attempt to save and retrieve a single stat.  I'm hoping that if I share the code from that here someone will be able to see what we're missing.

There are two scenes in the project.  The first one just signs the player in and then loads the second scene.  We know that part is working properly because the PlayerAuthentication prefab displays the correct Gamertag.  The second scene attempts runs through three things.  It attempts to retrieve a stat, then save a stat, then save a stat using an alternate method.  Evey time we run it the "retrieve stat" part cannot find a stat.  Here is the code.

`using Microsoft.Xbox.Services;
using Microsoft.Xbox.Services.Client;
using Microsoft.Xbox.Services.Statistics.Manager;
using System.Collections;
using UnityEngine;
using UnityEngine.UI;

public class MainTestScript : MonoBehaviour {

    public Text message;
    public IntegerStat Stat;
    XboxLiveUser user;
    StatisticValue statValue;
    bool started = false;

    private void Start()
    {
        StartCoroutine("RunScript");
    }

    IEnumerator RunScript()
    {
        yield return new WaitForSeconds(5f);
        GetUser();
        yield return new WaitForSeconds(10f);
        GetStat();
        yield return new WaitForSeconds(10f);
        SetStat();
        yield return new WaitForSeconds(50f);
        SetStatAnotherWay();
        yield return null;
    }

    void GetUser()
    {
        user = SignInManager.Instance.GetPlayer(1);
        message.text += "\n" + "User = " +  user.Gamertag.ToString();
    }

    void GetStat()
    {
        try
        {
            statValue = XboxLive.Instance.StatsManager.GetStatistic(user, "test");
        }
        catch
        {
            message.text += "\n" + "Unable to get stat from Xbox Live";
        }
        if (statValue != null)
        {
            message.text += "\n" + "Pulled score of " +  statValue.AsInteger.ToString() + " from Xbox Live.";
        }
    }

    void SetStat()
    {
        Stat.Value = 321;
        StatsManagerComponent.Instance.RequestFlushToService(user, false);
        message.text += "\n" + "Set IntegerStat to 321 and Flushed";
    }

    void SetStatAnotherWay()
    {
        XboxLive.Instance.StatsManager.SetStatisticIntegerData(user, "test", 123);
        message.text += "\n" + "Set stat to 123 directly and Flushed";
    }

}
`

We have a PlayerAuthentication Prefab in this scene that correctly displays the gamertag of the signed in user.  We have a text field called message (obvious by the code snippet) that correctly displays the gamertag as well.  Either we are not correctly saving the stat, or we are not correctly loading the stat.  Since we also have also tried using the Leaderboard prefab in another scene (hooked up to a correctly filled out IntegerStat prefab) and it shows as blank we're guessing it's the saving part we aren't getting.  

Here are some things we have already thought of:

We configured the stat with ID test (all lowercase) in Dev Center and pressed pressed Test (multiple times all with a Success message).

Our IntergerStat prefab has an ID of test (all lowercase).  We triple checked.

Our project is successfully Associated with our Game through Visual Studio.  

We've tried this on PC and Xbox One.  We've tried it with Debug projects as well as fully packaged projects.  

Any ideas?

**Comment by rakhad** [comment]
The code looks right. While the GetStatistic method is great for testing, you shouldn't use it as your production way of retrieving stat info. Featured Stats is not meant to be the source of truth and it is up to the game to ensure that the values that Featured Stats have are correct.

Some additional things you can check:
1) Ensure the user is added to StatsManager
2) Ensure that StatsManager.DoWork() method is being called regularly and that the LocalUserAdded event has been returned
3) Check that StatsManager.DoWork() returns a StatUpdateComplete event 

In addition, can you capture a fiddler trace of retrieving/setting the stat?

**Comment by AndyDnn** [comment]
Thanks rakhad. 

Thank you very much for your reply and for all your help.  We truly appreciate it.  

I would like to check on the three things you suggested and then capture a fiddler trace, but I don't have the slightest idea what any of what you said means.  I'm trying to comprehend it, but to me it's like watching the last 5 minutes of an episode of Star Trek: Voyager where they shout out a bunch of technobabble that magically solves all their problems.  Here is my interpretation of what I just read.

1. You're saying an XboxLiveUser has to be added to something called a StatsManager?  Is this something I was supposed to me doing in code?  I have no idea what this means let alone how to verify it's happening.  I would google it, but I already read through every piece of documentation on the Xbox Live Unity Prefab and Xbox Leaderboards/Stats many times and I didn't see it explained anywhere so there's literally nothing I can do.  

2. I have no idea what the StatsManager.DoWork() method is, if I'm supposed to be calling it myself somehow, or how I would verify it's working.  Also, there's a LocalUserAdded event?

3, Again, no idea what StatsManager.DoWork() even is.

4. There is absolutely no chance of any of us learning what Fiddler is in order to figure out why Stats/Leaderboards aren't working.  

At this point our goal is simple.  There is clearly correct code to "try" to save and load stats.  We just want to know what it is.  If you're saying our code looks like it should have worked, then I guess that means we already have it.  If we're already trying the correct code and it isn't working then we'll simply have to conclude that Xbox Leaderboards and Stats through the Prefab are not working correctly and give up on putting them in our game for now.  If anyone sees a flaw in our code and would suggest trying something different please let us know.

Again, sincere thank you to rakhad and to anyone else who decides to contribute a suggestion.

**Comment by BionicPony** [comment]
Hi Andy.  
   I have one idea what might be causing your problem. The Xbox Live Sandboxes are absurdly unreliable. It could be that your script would work fine in a published game but is failing in the test environment. I don't see a good way to test that theory. I suppose you could finish your game using the save code you showed here and hope for the best but that doesn't seem like a good idea.  I would tell you what I think of your code but to be honest I have never been able to get leaderboards working on Xbox either. When I saw your post I tried again over the weekend and had no luck. The only feedback I can give you is that I tried your exact script you posted and it didn't work for me either. 
   I do have one question for Ramsey Khadder if he sees this again. Of the two save stat methods Andy posted which one looks correct? SetStat() or SetStatAnotherWay()? I'll probably keep experimenting and trying to get it to work but knowing which one of those looks correct to you would be a big help. Also, is that method the ONLY code that needs to be called in order to save a stat or is there a missing step?  If I can find a solution during my testing I'll post it here for Andy and anyone else having similar issues.

**Comment by BionicPony** [comment]
Update:  We haven't been able to get stats to save but we can confirm a few things.  First, the problem is not related to the sandbox environment.  We added a High Score stat to a shipped game and it will not save there either.  Here is the code we tried.

`  stat.Value = score;
        XboxLive.Instance.StatsManager.SetStatisticIntegerData(user, "score", score);
        StatsManagerComponent.Instance.RequestFlushToService(user, true);`

That is the actual code we run to try to save the stat.  We haven't included any other code since it is not relevant.  Through other code and testing we are certain of a few things:

The variable "Stat" has been correctly linked to an IntegerStat prefab.  
A use is definitely signed in.
the variable "score" is definitely higher than zero in all of our tests.

The only logical conclusion is that setting an IntergerStat to a value and then calling `  StatsManagerComponent.Instance.RequestFlushToService(user, true);`
is clearly not the correct way to try to save a stat using the Xbox Live Unity Plugin.  There is something missing.  It is a shame that there are no instructions provided on how to use the plugin as some simple instructions would have eliminated this problem.  Instead we are left with only a monolithic example to try to reverse engineer if we want to try to decipher what code we're supposed to call in order to actually use the prefabs in the plugin, then do a series of trial and error research to test out theories.  This is frankly ridiculous.

If anyone from Microsoft sees this, how about just posting the actual line or two of code that we're supposed to be calling to save a stat using your plugin so that we, Andy, and anyone else who finds this thread can actually use it?

If my tone sounds harsh, I apologize, but I am very frustrated by this.  I've sunk a ridiculous amount of hours into trying to get this to work.  For comparison's sake, this took us 5 minutes to get up and running on every other console we've developed for, and we develop for all of them.

**Comment by rakhad** [comment]
Sorry for the delay.

Something interesting seems to be going on as you are calling the right methods. I will do a deep dive and report my findings by Monday. Thank you for your patience.

**Comment by BionicPony** [comment]
I worked on this over the weekend, but the only thing I have to report that could be any useful is that I tried to get it to work with a much older version of the Xbox Live Unity Plugin that I still have and I couldn't get it to work there either.  Then again that version was from the launch of the Creator's Program so it was so wildly different I might not have been trying the correct things.

I know for sure that it's a stat saving issue not a stat recall issue because the game "Space Command" which is in the store tries the code I've posted to save at multiple points in the game.  It's been selling well but zero people have a stat to report on their game hubs and the leaderboard is empty.

At least I have that game tracking people's personal high scores in local memory, so if I ever get this working I can add proper leaderboards.

**Comment by rakhad** [comment]
How do you handle sign in? Are you using the PlayerAuthentication prefab, using the SignInManager, or writing your own code?

**Comment by BionicPony** [comment]
We have Stats and Leaderboards working now thanks to your example.  We even updated one of our published games and it has working Leaderboards now (thank you SOO much).  We plan on walking back some of the changes and testing at each point to find exactly what the problem was and then posting that result here (even if the thread is closed) to help the person who originally opened this thread and anyone who finds it in the future.  Since we haven't heard from the original poster in 10 days it might be worth closing the thread if he doesn't post something soon.

**Comment by rakhad** [comment]
Glad to hear its working. 

By default stats are written to the service in 5 minute intervals to reduce load on the service. This might have been why stats weren't showing up. Flushing with a low priority takes all the pending stat writes and attempts to write them to the service on a 30 second interval. Flushing with a high priority takes all the pending stat writes and writes them immediately; however, flushing with high priority can only be called once every 30 seconds. You should flush with low priority when you finish a level and you should flush with high priority when the game is closing. I've included a few examples below.

Context
  1) Stats manager is initialized, the default 5 minute timer begins now, the priority 30 second time begins now 
  2) Level starts
  3) 30 seconds go by
  4) Level finishes
  5) Write stat to IntegerStat.Value
  6) Next level starts
  7) 1 minute go by
  8) Level finishes
  9) Write stat to IntegerStat.Value
10) 45 seconds go by

At this point we have 2 pending stat writes and 2 minutes and 15 seconds have elapsed

Default Behavior 
- There is 2 minutes and 45 seconds left on the timer before stats manager automatically uploads the pending 2 stat writes. If you attempt to get the stat value or get a leaderboard during that 2 minutes and 45 seconds then the value returned by the service will not reflect the pending writes.

Low Priority Flush
- Lets say we make a low priority flush after step 10.
- All pending stat writes are moved from the default queue to the priority queue. There is 15 seconds left on the priority timer before stats manager automatically uploads the pending 2 priority stat writes. If you attempt to get the stat value or get a leaderboard during that 15 seconds then the value returned by the service will not reflect the pending writes.

High Priority Flush
- Lets say we are using the default behavior and our 2 pending stat writes are on the default queue. Lets also say the user decides to close the game and Unity gives us the opportunity to handle the game exiting. In the Unity callback we would call a high priority flush.
- Stat manager will attempt to immediately write all pending stat writes to the service assuming there has not been another high priority flush in the last 30 seconds. In this case there hasn't been and the stat writes are immediately written to the service.
- Note if there has been a high priority flush in the last 30 seconds then the pending stat writes will be moved to the priority queue and will have to wait until the next time the 30 second priority timer completes.

More detail can be found in the stat manager [docs](https://docs.microsoft.com/en-us/windows/uwp/xbox-live/leaderboards-and-stats-2017/stats2017).
